integration-tests/src/tests/client/resharding.rs

```diff
@@ -1477,6 +1488,141 @@ fn test_shard_layout_upgrade_cross_contract_calls_v3_seed_44() {
+    let mut init_txs = vec![];
+    for account_id in &contract_accounts {
+        let signer =
+            InMemorySigner::from_seed(account_id.clone(), KeyType::ED25519, account_id.as_ref());
+        let actions = vec![Action::DeployContract(DeployContractAction {
+            code: near_test_contracts::nightly_rs_contract().to_vec(),
+        })];
+        let init_tx = SignedTransaction::from_actions(
+            1,
+            account_id.clone(),
+            account_id.clone(),
+            &signer,
+            actions,
+            genesis_hash,
+        );
+        init_txs.push(init_tx);
+    }
+    test_env.set_init_tx(init_txs);
+
+    let mut yield_tx_hashes = vec![];
+    let mut nonce = 100;
+
+    // In these tests we set the epoch length equal to the yield timeout length.
+    assert!(
+        epoch_length
+            == RuntimeConfig::test().wasm_config.limit_config.yield_timeout_length_in_blocks,
+    );
+    // Add transactions invoking promise_yield_create near the epoch boundaries.
+    for height in [
+        epoch_length - 2, // create in first epoch, trigger timeout during resharding epoch
+        epoch_length - 1,
+        epoch_length,
+        2 * epoch_length - 2, // create during resharding epoch, trigger timeout on upgraded layout
+        2 * epoch_length - 1,
+        2 * epoch_length,
+        3 * epoch_length - 2, // both the create and the timeout will occur on upgraded layout
+        3 * epoch_length - 1,
+        3 * epoch_length,
+    ] {
+        let txs = contract_accounts
+            .iter()
+            .map(|account_id| {
+                let signer = InMemorySigner::from_seed(
+                    account_id.clone(),
+                    KeyType::ED25519,
+                    account_id.as_ref(),
+                );
+                nonce += 1;
+                let tx = SignedTransaction::from_actions(
+                    nonce,
+                    account_id.clone(),
+                    account_id.clone(),
+                    &signer,
+                    vec![Action::FunctionCall(Box::new(FunctionCallAction {
+                        method_name: "call_yield_create_return_promise".to_string(),
+                        args: vec![],
+                        gas: GAS_1,
+                        deposit: 0,
+                    }))],
+                    genesis_hash,
+                );
+                yield_tx_hashes.push(tx.get_hash());
+                tx
+            })
+            .collect();
+        test_env.set_tx_at_height(height, txs);
+    }
+
+    yield_tx_hashes
+}
+
+// Test delivery of promise yield timeouts
+fn test_shard_layout_upgrade_promise_yield_impl(resharding_type: ReshardingType, rng_seed: u64) {
+    init_test_logger();
+
+    // setup
+    let epoch_length =
+        RuntimeConfig::test().wasm_config.limit_config.yield_timeout_length_in_blocks;
+    let genesis_protocol_version = get_genesis_protocol_version(&resharding_type);
```

Comment:
I made some progress adding the requested unit test, in which promise yield is started before shard layout switch and finished after. One tricky thing is that the resharding tests work with some specific protocol versions `SimpleNightshade`, `SimpleNightshadeV2`, `SimpleNightshadeV3`, none of which should have yield execution enabled.

Any suggestions on the nicest way to resolve this?

Reply:
It is worth noting there are other implications as well. `ProtocolFeature::ComputeCosts` came after `SimpleNightshade`. Rolling back the protocol version does not roll back the runtime config specifying gas and compute costs. Hence any contracts invoking certain host functions (e.g `storage_write`) are incompatible with V1 resharding tests (the validators panic [here](https://github.com/near/nearcore/blob/d92c69d6cd8cd0c2b1a52104270ae58d34fa5cf6/runtime/runtime/src/lib.rs#L1447-L1449)).

Reply:
> One tricky thing is that the resharding tests work with some specific protocol versions SimpleNightshade, SimpleNightshadeV2, SimpleNightshadeV3, none of which should have yield execution enabled.

> Any suggestions on the nicest way to resolve this? 

We can add a new resharding to the nightly protocol after yield resume and rely on the nightly part of CI to cover this case.

Reply:
Ah actually that wouldn't work because nightly also contains stateless validation and it's not compatible with resharding, sorry! 

I still would want to have the test checked in since it's good stuff and will be useful for the next resharding, whenever it may be. I would suggest either of two options:
* Set it up manually by disabling stateless validation and run it just once to see that it works fine. Then undo the one-off setup, add a TODO, disable or ignore the test and merge it to master. 
* Set it up with a new custom rust feature e.g. `simple_nightshade_v4_protocol` that would fall somewhere between stable (65) and stateless validation (80) protocol versions. It won't be run by CI and we will need to get rid of it before stabilizing stateless validation but at least it will be possible to run it with --features simple_nightshade_v4_protocol.

Reply:
OK that makes sense, I took a stab at adding a nightly resharding and observed that it fails iff stateless validation is enabled. Will see what I can do (leaning towards something along the lines of the second option).

Reply:
All right, please ping me or refresh review when ready.

Reply:
I ended up doing something slightly tricky in epoch manager which allows us to include this test in CI, PTAL. 

It remains to gate yield execution behind the new `ProtocolFeature::YieldExecution` but this PR has grown large enough already; I'll send a separate one. Technically YieldExecution currently requires version >= 139 because the host functions are enabled in the runtime config snapshot #139. But the resharding tests don't switch the runtime config when they change protocol versions so everything works right now.